docs/polkadot/implementors-guide/guide.md

```diff
@@ -0,0 +1,660 @@
+  * Candidates that have just become available should apply any pending code upgrades based on the relay-parent they are targeting and should schedule any upcoming pending code upgrades.
+
+candidates entering the "pending approval" state:
+  * Apply fees (TODO: not sure if fees are actually used, we don't seem to need 'em for XCMP)
+  * Apply pending code upgrade, if any.
+  * Schedule a new pending code upgrade if the candidate specifies any. (there is a race condition here: part of the configuration is "how long should it take before pending code changes are applied". This value is computed based on the relay-parent that was used at the point when the candidate was about to be included in the relay chain. This is potentially a few blocks later than that, as it can take some time for a candidate to become fully available. We need to ensure that the code upgrade is scheduled with the same delay as was expected when the code upgrade was signaled. The easiest thing to do is to make sure the `pending_code_delay` is passed through the entire availability pipeline).
+  * Schedule Upwards messages - messages from the parachain to the relay chain.
+
+process new backed candidates:
+  * ensure that only one candidate is backed for each parachain or parathread
+  * ensure that the parachain or parathread of the candidate was scheduled and does not currently have a block pending availability. 
+  * check the backing of the candidate. 
+  * move to "pending approval" state. (pass along any configuration information that is liable to change)
+
+misbehavior reports and secondary checks:
+  * Secondary checks will also be submitted within the block. This may lead to slashing as a secondary check period ends. We want to catch and punish for the cases of misbehavior that violate the protocol and put its security at risk. One of such cases is submitting conflicting votes on the same `CandidateReceipt`.Other examples include violations to AnV protocol or equivocations in finality. Misbehavior handling is implemented in
+    * Runtime as an entry point.
+    * Code in the Node that assists submitting misbehavior reports.
+
+finalization: (not finality)
+  * ensure that required updates (bitfields and backed candidates) occurred within the block.
+  * update scheduling metadata based on parachains that had blocks included or not. for instance, parathreads where the auction-winning collator didn't get a chance to include its block should be allowed to retry a couple of times.
+
+Availability bitfields must go in before parachain candidates, otherwise there would be a minimum of 1 relay chain block between blocks of the same parachain. As such, it's best for them to go into the same extrinsic.
```

Comment:
We could restrict the block structure elsewhere too instead, right?

Reply:
Missed this until now. could you elaborate on this comment? I'm not sure what you mean by it

Reply:
We could enforce extrinsics appear in some specified order inside the block, no?  

We read some series of `CandidateBacked`, `AvailabilityVote`, and `CandidateAvailable` extrinsics.  I think `CandidateAvailable` refers to a specific `CandidateReceipt` but otherwise only states that enough `AvailabilityVote`s happened to trigger tallies during block verification.  We cannot avoid `AvailabilityVote`s each having their own relay chain parent, possibly the chain head, but not the current block.  We mask out any new `CandidateBacked` transactions between the votes parent and the chain head when processing, so if we process any `CandidateBacked` before a corresponding `CandidateAvailable` then we remove that bit form the mask, and block that candidate being available.  It sounds okay, although might enable censorship of parathreads, so we could invalidate the block if any `AvailabilityVote`s or `CandidateAvailable` have parents that precede the current `CandidateBacked`.  In effect, this imposes the desired block structure.  

Any threads processing `AvailabilityVote`s for the same candidate must be joined before its `CandidateAvailable` I guess.  I think `CandidateBacked` could safely run inside another thread even under the weaker memory model.

Reply:
It doesn't seem like we need a `CandidateAvailable` extrinsic since availability is computed implicitly on processing bitfields. Furthermore we want to process all bitfields at once since interleaving sounds complicated. I'm still in favor of bundling them since it doesn't seem we lose much by it